Fix #166: generateKeyApi not flexible enough.

bblfish opened the pull request:
With these changes it is possible to code the [WebCrypto API signing example](http://www.w3.org/TR/WebCryptoAPI/#examples-signing) in the same style as the example fixing #166 

```scala
    GlobalCrypto.crypto.subtle.generateKey(
      RsaHashedKeyGenParams(
        2048,
        new Uint8Array(js.Array[Short](1,0,1)), //65537
        HashAlgorithm.`SHA-256`),
      false,
      js.Array(KeyUsages.sign,KeyUsages.verify)
    )
```

bblfish commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -257,17 +278,17 @@ object RsaKeyGenParams {
-    js.Dynamic.literal(name = name, modulusLength = modulusLength.toDouble,
+  def apply( modulusLength: Long,
+      publicExponent: BigInteger, hash: HashAlgorithmIdentifier): RsaHashedKeyGenParams = {
+    js.Dynamic.literal(name = "RSASSA-PKCS1-v1_5", modulusLength = modulusLength.toDouble,
```
note here the removal of the `name` variable, and its replacement in body with a constant.  If you agree with this, then it one should do this with the other Algorithms. It would be really nice to be able to get out of this pattern where names are passed as strings with no type safety.

mseddon replied:
I absolutely agree in the normal case this should be discouraged, but perhaps we should at least expose an overloaded apply with a name in case one is using an algorithm not covered by the library?

bblfish replied:
Is it actually possible to do that?

mseddon replied:
Yes, this is absolutely correct, actually.  Let's try and get rid of these magic constants where possible.

mseddon replied:
Bleh- sorry, deleted and reposted my comment by mistake and now we're all out of sync.  I think you're doing the right thing here, so `RsaHashedKeyGenParams` is correct (no need for the name parameter).

One thing though- the `BigInteger` exponent might want to be overloaded.  For example I might legitimately have a `Uint8Array` from somewhere else and want to instance an `RsaHashedKeyGenParams`, so while the `BigInteger` is definitely convenient, the lower level `Uint8Array` should be kept, too.

bblfish replied:
I have a feeling that you have a better mental model of the cryptography algorithms than I do. You could either submit a different pull request that aligns the others with this pattern, or make a pull request to my branch with those changes, that I can accept and that will then appear here. I don't feel very confident to do this with the other algorithms ( having ever only really used RSA ).

mseddon replied:
Okay, that's no problem. There isn't a lot of heavy crypto api users (I believe the list is currently, us. :)), and I still want to go through this api and see if I can strengthen all these Promise[Any]'s.  I will touch most of these classes anyway in that PR.

mseddon commented:
@sjrd already reviewed in #168, LGTM. WDYT?

bblfish commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -290,14 +311,16 @@ object RsaKeyAlgorithm {
 @js.native
 trait RsaHashedKeyAlgorithm extends RsaKeyAlgorithm {
-  var hash: KeyAlgorithm = js.native
+  var hash: HashAlgorithm = js.native
```
I fixed this automatically. I wonder if they may not have a mistake in the spec though.
http://www.w3.org/TR/WebCryptoAPI/#RsaHashedKeyAlgorithm-dictionary

mseddon replied:
I had a gut feeling about that line, too.  With the spec as it currently stands, `HashAlgorithm` is definitely an appropriate type.  We might have to weaken that due to future changes, but I think this for now is closer to the semantics.

bblfish replied:
On monday I can ask on a number of w3c channels to see if we have misunderstood or if it is a bug.

mseddon replied:
That would be good.  I suspect it's just not as tight as it could be, but if we're off fundamentally, that would be nice to know :)

bblfish commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -305,14 +328,14 @@ object RsaHashedKeyAlgorithm {
-trait RsaHashedImportParams extends js.Object {
-  var hash: String = js.native
+trait RsaHashedImportParams extends js.Object with HashAlgorithm {
+  var hash: HashAlgorithm = js.native
```
I should probably change this to `HashAlgorithmIdentifier` as that is what is in the spec http://www.w3.org/TR/WebCryptoAPI/#RsaHashedImportParams-dictionary
( perhaps we should put links for each of these to the spec )
I only added a HashAlgorithmIdentifier in the last round of changes, and must have missed this one.

mseddon replied:
Actually, also `HashAlgorithm` extends `js.Object`, so just `extends HashAlgorithm`

mseddon commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -257,17 +278,17 @@ object RsaKeyGenParams {
 @js.native
 trait RsaHashedKeyGenParams extends RsaKeyGenParams {
-  var hash: String = js.native
+  var hash: HashAlgorithm = js.native
```
http://www.w3.org/TR/WebCryptoAPI/#RsaHashedKeyGenParams-dictionary

I think this one wants to be `HashAlgorithmIdentifier`, too.

bblfish commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -305,14 +328,14 @@ object RsaHashedKeyAlgorithm {
   @inline
-  def apply(hash: String): RsaHashedImportParams =
+  def apply(hash: HashAlgorithm): RsaHashedImportParams =
     js.Dynamic.literal(hash = hash).asInstanceOf[RsaHashedImportParams]
```
Note that when I change line 339 to `hash: HashAlgorithIdentifier` then I need to change the line below to `hash.asInsanceOf[js.Dynamic]` or it won't compile. You'll see this elswhere too.

mseddon replied:
Huh. Yes, I see, seems to be that Scala.js doesn't think `HashAlgorithm | String` extends `js.Any`. This appears to be a deficiency in the union types. @sjrd thoughts?

sjrd replied:
Well since `String` is not a subtype of `js.Any`, it's normal that a union of String with something else is not either. Casting is the way to go, here, but cast to `js.Any`, not `js.Dynamic`

mseddon replied:
Ah, right, of course that won't work.  Not a deficiency in union types, a deficiency in my thinking. :)

bblfish commented:
I pushed a little update to this PR with the following improvements:

1. changed `KeyUsages` to `KeyUsage` as per w3c spec
2. added some initial ScalaDoc pointing to w3c spec, which is the standard after all
3. fixed the `.asInstanceOf[js.Dynamic]` to `asInstanceOf[js.Any]`

That should be good enough for this PR I think.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -1,9 +1,15 @@
 package org.scalajs.dom.crypto
 
-import scalajs.js
+import java.math.{BigInteger => _}
```
Why is this needed? `java.math.*` is not imported by default.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -43,6 +49,21 @@ trait Algorithm extends js.Object {
+
+}
+
+
```
Double blank line.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -43,6 +49,21 @@ trait Algorithm extends js.Object {
+
+object HashAlgorithm {
+
+  private def named(name: String): HashAlgorithm = literal(name=name).asInstanceOf[HashAlgorithm]
```
Try to keep line length under 80 characters. You can easily put the body on a separate line.
Also, add spaces around `=`.

sjrd replied:
In fact, this does not seem right. `HashAlgorithm` extends `Algorithm`, which has a `name` field. But a `String` does not have a `name` field. This cast is misleading. What is happening, here?

bblfish replied:
This is `js.Dynamic.literal(name=name)` which I think produces 
```JSON
{
  "name": "$name"
}
```

sjrd replied:
Oh right. I completely misread the whole thing. Never mind that.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -1,9 +1,15 @@
 import scala.scalajs.js.typedarray.ArrayBufferView
 
+
+
```
Triple blank line.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/package.scala:
```diff
@@ -2,12 +2,16 @@ package org.scalajs.dom
 
   type BigInteger = Uint8Array
+  type AlgorithmIdentifier = KeyAlgorithm | String
+  type HashAlgorithmIdentifier = HashAlgorithm | String
```
According to http://www.w3.org/TR/WebCryptoAPI/#algorithm-dictionary, `HashAlgorithmIdentifier` is the same as `AlgorithmIdentifier`. Why is it not the case here?

bblfish replied:
I was thinking that one should use the type system of the JS spec as much as possible, but improve on it where it makes sense. JS developers and spec writers are perhaps not so used to working with a strong type system, and may be missing out from time to time on some nice semantic differences they could gain by being clearer.

For example their definition is this:

```
typedef (object or DOMString) AlgorithmIdentifier;
```

It seems a bit weird that they use `object` rather than `Algorithm` there for example. 

In this case there seems to be a clear distinction between a `KeyAlgorithm` and a `HashAlgorithm`. Even though they have the same structure ( they are both objects with a "name" field ), they seem to be non overlapping classes. If this is right then making the distinction in the type system would make it easier for the user to find the right types to use.

I could leave a note explaining the difference in the code, so that if at some point we do find it justified, then it would be easy to see where the difference came from.

sjrd replied:
OK

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -725,16 +758,63 @@ trait Pbkdf2Params extends Algorithm {
+  val deriveKey  = "deriveKey".asInstanceOf[KeyUsage]
+  val deriveBits = "deriveBits".asInstanceOf[KeyUsage]
+  val wrapKey    = "wrapKey".asInstanceOf[KeyUsage]
+  val unwrapKey  = "unwrapKey".asInstanceOf[KeyUsage]
```
Do not vertically align `=`s.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -725,16 +758,63 @@ trait Pbkdf2Params extends Algorithm {
+ * See [[http://www.w3.org/TR/WebCryptoAPI/#cryptokey-interface ¶ 13. CryptoKey Interface]] of w3c spec
+ */
+@js.native
+trait KeyUsage  extends js.Any
```
Double space before `extends`.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -725,16 +758,63 @@ trait Pbkdf2Params extends Algorithm {
+object KeyType {
+  val public    = "public".asInstanceOf[KeyType]
+  val `private` = "private".asInstanceOf[KeyType]
+  val secret    = "secret".asInstanceOf[KeyType]
```
Do not vertically align.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -725,16 +758,63 @@ trait Pbkdf2Params extends Algorithm {
+  /** The DER encoding of the SubjectPublicKeyInfo structure from RFC 5280. */
+  val spki  = "spki".asInstanceOf[KeyFormat]
+  /** The key is a JsonWebKey dictionary encoded as a JavaScript object */
+  val jwk   = "jwk".asInstanceOf[KeyFormat]
```
Do not vertically align.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -725,16 +758,63 @@ trait Pbkdf2Params extends Algorithm {
+  /** The key is a JsonWebKey dictionary encoded as a JavaScript object */
+  val jwk   = "jwk".asInstanceOf[KeyFormat]
+
+}
```
Missing EOL at EOF

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -725,16 +758,63 @@ trait Pbkdf2Params extends Algorithm {
+trait KeyUsage  extends js.Any
+
+object KeyUsage {
+  val encrypt    = "encrypt".asInstanceOf[KeyUsage]
```
Maybe these constants should start with an uppercase in Scala? I am not sure, here. It might be better to stick to the identifier version of the string they represent.

bblfish replied:
ok. I'll stick to it.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/package.scala:
```diff
@@ -2,12 +2,16 @@ package org.scalajs.dom
 
 import scala.scalajs.js
 import scala.scalajs.js.typedarray.{ArrayBufferView, ArrayBuffer, Uint8Array}
+import scala.scalajs.js.`|`
```
The backsticks are unnecessary.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -288,32 +311,42 @@ object RsaKeyAlgorithm {
 }
 
+
+
```
Triple blank line.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -257,17 +280,17 @@ object RsaKeyGenParams {
-  def apply(name: String, modulusLength: Long,
-      publicExponent: BigInteger, hash: String): RsaHashedKeyGenParams = {
-    js.Dynamic.literal(name = name, modulusLength = modulusLength.toDouble,
+  def apply( modulusLength: Long,
```
Spurious space after `(`

bblfish commented:
* added documentation pointing to the W3C spec.
* pointed out where we go a bit beyond the w3c spec in terms of type safety ( so that if I made a mistake it's easy for someone to work out where it may have come from )
* removed the white space
* changed `literal(...)` to `js.Dynamic.literal(...)` to keep code consistent with other uses

Hope that's good enough for this PR.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -112,106 +142,145 @@ trait JsonWebKey extends js.Object {
-  def verify(algorithm: String, key: CryptoKey, signature: BufferSource,
-      data: BufferSource): Promise[js.Any] = js.native
+  def verify(algorithm: KeyAlgorithmIdentifier, key: CryptoKey,
+             signature: BufferSource, data: BufferSource): Promise[js.Any] = js.native
```
Indentation issue.

sjrd commented on src/main/scala/org/scalajs/dom/crypto/Crypto.scala:
```diff
@@ -112,106 +142,145 @@ trait JsonWebKey extends js.Object {
+   *
+   * We are being a bit more precise than the official definition by requiring a KeyAlgorithmIdentifier
+   * rather than an AlgorithmIdentifier.
+
```
Spurious blank line.

bblfish replied:
oops. I just realised which line you meant was spurious.

sjrd commented:
That's all. Could you also squash all the commits into a single commit, please?

bblfish commented:
done

sjrd commented:
LGTM

bblfish commented:
Ok should do now. I removed the supeflous blank line.

bblfish commented:
Ok should do now, I really removed the blank line!

mseddon commented:
@bblfish Thanks!
